### PR TITLE
Remove atexit() calls and implement a diablo_deinit() function instea…

### PR DESCRIPTION
**Pull request by vanfanel**
…d that cleans up every subsystem if it has been init before.

Fixes https://github.com/diasurgical/devilutionX/issues/689 and possibly other atexit-related problems.

### CODE REVIEW

File: Source/diablo.h [hunk 1]
```diff
@@ -26,6 +26,9 @@ extern int DebugMonsters[10];
 extern BOOLEAN cineflag;
 extern int force_redraw;
 extern BOOL visiondebug;
+/* These are defined in fonts.h */ 
+extern BOOL was_ttf_init;
```
Comment (AJenbo) [inline Source/diablo.h:30]: was_fonts_init

File: Source/diablo.cpp [hunk 2]
```diff
@@ -270,13 +276,13 @@ void diablo_init()
 
 	SFileEnableDirectAccess(TRUE);
 	init_archives();
-	atexit(init_cleanup);
+	was_diablo_init = true;
```
Comment (AJenbo) [inline Source/diablo.cpp:279]: was_archives_init

File: Source/init.h [hunk 3]
```diff
@@ -27,6 +27,9 @@ LRESULT MainWndProc(HWND hWnd, UINT Msg, WPARAM wParam, LPARAM lParam);
 void init_activate_window(HWND hWnd, BOOL bActive);
 WNDPROC SetWindowProc(WNDPROC NewProc);
 
+extern BOOL was_dx_init;     // defined in dx.cpp
```
Comment (AJenbo) [inline Source/init.h:30]: Window or graphics

File: SourceX/DiabloUI/fonts.cpp [hunk 4]
```diff
@@ -5,6 +5,8 @@ namespace dvl {
 TTF_Font *font = nullptr;
 BYTE *FontTables[4];
 Art ArtFonts[4][2];
+/* This is so we know ttf has been init when we get to the diablo_deinit() function */
```
Comment (AJenbo) [inline SourceX/DiabloUI/fonts.cpp:8]: /**

So that it is picked up as docs

File: SourceX/DiabloUI/fonts.cpp [hunk 5]
```diff
@@ -74,4 +76,8 @@ void UnloadTtfFont() {
 	font = nullptr;
 }
 
+void DeinitTtf() {
```
Comment (AJenbo) [inline SourceX/DiabloUI/fonts.cpp:79]: FontsCleanup

File: SourceX/display.cpp [hunk 6]
```diff
@@ -19,6 +19,8 @@
 
 namespace dvl {
 
+extern BOOL was_dx_init; // defined in dx.cpp
```
Comment (AJenbo) [inline SourceX/display.cpp:22]: Change comments to documentation.

File: Source/diablo.h (hunk 1 above)
Comment (vanfanel) [reply Source/diablo.h]: Done!

File: Source/diablo.cpp (hunk 2 above)
Comment (vanfanel) [reply Source/diablo.cpp]: Done!

File: Source/init.h (hunk 3 above)
Comment (vanfanel) [reply Source/init.h]: Done!

File: SourceX/DiabloUI/fonts.cpp (hunk 4 above)
Comment (vanfanel) [reply SourceX/DiabloUI/fonts.cpp]: Done!

File: SourceX/DiabloUI/fonts.cpp (hunk 5 above)
Comment (vanfanel) [reply SourceX/DiabloUI/fonts.cpp]: Done!

File: SourceX/display.cpp (hunk 6 above)
Comment (vanfanel) [reply SourceX/display.cpp]: Done!

**Comment by vanfanel** [comment]
@AJenbo I have done all asked changes. I hope it can be merged now :)

**Comment by vanfanel** [comment]
Thanks @AJenbo ! ^_^"

**Pull request merged**